### ISSUE TITLE
Running direnv from script

### ISSUE THREAD

**Issue by rbutola** [direnv/direnv#262]
Hi,

Problem:  "direnv hook bash" does not work from shell script. 
Root Cause: This happens because "direnv hook" command appends "_direnv_hook" function to PROMPT_COMMAND. When runnning a normal shell script, command prompt will be presented only when script ends. Therefore any direnv related action would happen effectively after the script has ended.
Temporary workaround: Following 3 commands achieve the desired result of exporting .envrc
eval "$(direnv hook bash)" # Introduces _direnv_hook function in shell
direnv allow # allows any direnv operation
_direnv_hook  # executes hook funtion - exports bash variables

Pls advice if this is the intended behavior.
Can I add a new command for hook to work from inside shell scripts.

**Comment by rbutola** [comment]
Sorry to disturb, just 1 point that came up when we discussed your
suggestion.
In non-interactive mode - how will the unloading part work if we use this
approach.

As I see the code, direnv export command is responsible for even unloading
part as well.
So on each directory change (cd to another directory) - do we explicitly
run eval "$(direnv export bash)" for unloading to work.



Thanks
Rahul
Mob: 8800740896

**Comment by zimbatm** [comment]
Yes that's it, execute `eval "$(direnv export bash)"` every time you want to change environment.

If you need to do it multiple times then it's possible to wrap the functionality in a bash function like this:
```sh
function direnv_cd() {
  cd "$1"
  eval "$(direnv export bash)"
}

direnv_cd some/dir
direnv_cd some/other/dir
```

**Comment by nicknovitski** [comment]
Sorry for the issue necromancy, but is `direnv exec` equivalent to the eval-and-export approach you describe?

Currently I reuse .envrcs in our CI by specifying that the build steps are run with an interactive shell, but I'm considering switching to a function like:
```
r() {
  direnv exec . "$@"
}
```

**Comment by lsiden** [comment]
To get this to work, I had to enable iterm2 shell integration only _after_ eval "$(direnv hook bash)".

The end of my .bashrc:

    # https://direnv.net/
    eval "$(direnv hook bash)"
    echo PRMOPT_COMMAND=$PROMPT_COMMAND

    if [ -f "${HOME}/.iterm2_shell_integration.bash" ]; then
      source "${HOME}/.iterm2_shell_integration.bash"
      export PS1='\[$(iterm2_prompt_mark)\]'$PS1
    fi

**Comment by dmayle** [comment]
I'm trying to understand this from the point of view of a script I want to install into a direnv and have it run from the direnv it's inside of.  Ideally, I could have two copies of the exact same script, installed in two different direnv's, and when calling each one, they would run from their respective direnv.

Right now, it looks like the required magic is:
```
#!/usr/bin/env bash
pushd $(dirname $(which $0))
eval "$(direnv export bash)"

echo Here
```
But when I run that, I get the following:
```
./myenv/test.sh: line 3: .: filename argument required
.: usage: . filename [arguments]
./myenv/test.sh: line 4: .: filename argument required
.: usage: . filename [arguments]
```

**Comment by arthurwolf** [comment]
I have a very similar issue, and I was wondering if anyone here has run into something similar. 

I need to run "direnv allow" before doing "make" ("wire-server github repo, docs folder/building).

When I do it manually, it works fine.

When I do "direnv allow" in a shell script, nothing happens (and "make" breaks the same way it does when running manually and I forget to type "direnv allow").

Any idea how to fix this?

Thanks so much.

**Comment by andy108369** [comment]
The main issue for me was that I've used `direnv hook bash` instead of `direnv export bash`.

Changing that made it work in my github action script I am working on.

```
eval "$(direnv export bash)"
```

I did not even need to set `direnv allow .`

**Comment by thenbe** [comment]
For non-interactive use, how would you activate an `.envrc` that is located in `some/path`?

Currently, I'm just starting a shell directly in `some/path` then running `eval "$(direnv export bash)"`. But I was wondering if it was possible to activate `some/path/.envrc` without having to set the directory to `some/path`.

**Comment by nicknovitski** [comment]
> For non-interactive use, how would you activate an `.envrc` that is located in `some/path`?
> 
> Currently, I'm just starting a shell directly in `some/path` then running `eval "$(direnv export bash)"`. But I was wondering if it was possible to activate `some/path/.envrc` without having to set the directory to `some/path`.

Yeah, with the `direnv exec ` command.

**Comment by zachdaniel** [comment]
Something feels off to me here. I don't want to do a blanket "direnv allow", I want to just do the `direnv export zsh` bit, and have it only use things that I've allowed manually in the past. Anything else feels like a security issue to me. Is there some way to do that? For `direnv export` to do anything at all for me I first have to do `direnv allow .`, which seems not ideal at all.

**Comment by zachdaniel** [comment]
I ended up doing this in my zshenv

```zsh
allow_dir=${XDG_DATA_HOME:-$HOME/.local}/share/direnv/allow
allow_paths=`ls $allow_dir/* | xargs cat | sort | uniq`
echo "$allow_paths" | while IFS= read -r allow_path; do
  dir_path=$(dirname "$allow_path")
  if [[ "$PWD" == "$dir_path"* ]]; then
    source $dir_path/.envrc
  fi
done

```
